### PR TITLE
metadata validator for smartspim

### LINKED ISSUE #547: metadata completeness check for SmartSPIM data
# User story
As a user, I want to ensure metadata is complete when uploading ephys data, so I can avoid uploading data assets with incomplete metadata.

# Acceptance criteria
- [x] Given a user creates a Metadata model, when the data_description is not None and the data_description.platform field is set to SmartSPIM, then an error will be raised if subject, procedures, acquisition, and/or instrument is None. (This can be accomplished through an extra model_validator in the Metadata class)
- [x] Given a user creates a Metadata model, when the data_description is not None and the data_description.platform field is set to SmartSPIM, then an error will be raised if procedures is missing Injection Materials

# Sprint Ready Checklist 
- [x] 1. Acceptance criteria defined 
- [x] 2. Team understands acceptance criteria 
- [x] 3. Team has defined solution / steps to satisfy acceptance criteria 
- [x] 4. Acceptance criteria is verifiable / testable 
- [x] 5. External / 3rd Party dependencies identified
- [x] 6. Ticket is prioritized and sized

# Notes
Does it make sense to move the upload job models into this repo?

## Original issue:
```
Internally collected `SmartSPIM` data are mature enough that we can require that metadata be complete.  

For `SmartSPIM`, this means:
- `data_description.json` exists
- `subject.json` exists for Allen Institute mice
- `procedures.json` exists for Allen Institute mice
- `acquisition.json` exists
- `instrument.json` exists

Acceptance criteria:
- `SmartSPIM` experiment type data cannot be uploaded if the above criteria are not true.
    - "Allen Institute mice" are those that do not have an `AK` prefix in their subject ID.
```

### PR DESCRIPTION
**Pull request by mekhlakapoor**
closes #547

### CODE REVIEW

File: src/aind_data_schema/core/metadata.py [hunk 1]
```diff
@@ -181,6 +181,25 @@ def validate_metadata(self):
...
+        ):
+            raise ValueError(
+                "Missing some metadata for SmartSpim. Requires subject, procedures, acquisition, and instrument."
+            )
+        if (
+            self.data_description
+            and self.data_description.platform == SmartSpim
+            and (self.procedures and getattr(self.procedures, "injection_materials", None) is None)
+        ):
+            raise ValueError("Procedures is missing injection materials.")
```
Comment (saskiad) [inline src/aind_data_schema/core/metadata.py:200]: do all SmartSpim procedures involve injections? Would it be better to hinge this warning on if there is an Injection procedure in the list of subject procedures?

Comment (mekhlakapoor) [reply src/aind_data_schema/core/metadata.py:200]: The injection materials check is based on the 2nd acceptance criteria here: https://github.com/AllenNeuralDynamics/aind-data-schema/issues/547 , but let me know if we should update that criteria.

File: src/aind_data_schema/core/metadata.py [hunk 2]
```diff
@@ -181,6 +181,25 @@ def validate_metadata(self):
...
+        return self
+
```
[the remark this answers is not in the document]
Comment (mekhlakapoor) [reply src/aind_data_schema/core/metadata.py:202]: Also was not in the acceptance criteria but I can add that if we want. Should we move this to discussions thread?

[the remark this answers is not in the document]
Comment (jtyoung84) [reply src/aind_data_schema/core/metadata.py:202]: I think it should be easy enough to add that check in this PR if we want that

File: src/aind_data_schema/core/metadata.py [hunk 3]
```diff
@@ -195,7 +221,13 @@ def validate_ecephys_metadata(self):
         if (
             self.data_description
             and self.data_description.platform == Ecephys
-            and (self.procedures and getattr(self.procedures, "injection_materials", None) is None)
+            and self.procedures
+            and any(
+                isinstance(surgery, BrainInjection) and getattr(surgery, "injection_materials", []) == []
```
Comment (saskiad) [inline src/aind_data_schema/core/metadata.py:226]: ditto

**Review by saskiad — CHANGES_REQUESTED** [review]
I think we're better off using `Injection` rather than `BrainInjection`. Otherwise I think it looks good.

**Review by saskiad — APPROVED** [review]

**Pull request merged**